src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,227 +9,316 @@ internal struct FileStatus
-            bool isDirectory)
-        {
-            status.InitiallyDirectory = isDirectory;
-            status._fileStatusInitialized = -1;
-        }
+        // Is a directory as of the last refresh
+        private bool _isDirectory;
 
-        internal void Invalidate() => _fileStatusInitialized = -1;
+        // Checks if the main path (without following symbolic links) has the hidden attribute
+        // Only call if Refresh has been successfully called at least once
+        internal bool HasHiddenFlag =>
+            (_mainCache.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN;
 
-        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        // Checks if the main path (without following symbolic links) has the read-only attribute
+        // Only call if Refresh has been successfully called at least once
+        private bool HasReadOnlyFlag
         {
-            EnsureStatInitialized(path, continueOnError);
+            get
+            {
 #if TARGET_BROWSER
-            const Interop.Sys.Permissions readBit = Interop.Sys.Permissions.S_IRUSR;
-            const Interop.Sys.Permissions writeBit = Interop.Sys.Permissions.S_IWUSR;
+                const Interop.Sys.Permissions readBit = Interop.Sys.Permissions.S_IRUSR;
+                const Interop.Sys.Permissions writeBit = Interop.Sys.Permissions.S_IWUSR;
 #else
-            Interop.Sys.Permissions readBit, writeBit;
+                Interop.Sys.Permissions readBit, writeBit;
 
-            if (_fileStatus.Uid == Interop.Sys.GetEUid())
-            {
-                // User effectively owns the file
-                readBit = Interop.Sys.Permissions.S_IRUSR;
-                writeBit = Interop.Sys.Permissions.S_IWUSR;
+                if (_mainCache.Uid == Interop.Sys.GetEUid())
+                {
+                    // User effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRUSR;
+                    writeBit = Interop.Sys.Permissions.S_IWUSR;
+                }
+                else if (_mainCache.Gid == Interop.Sys.GetEGid())
+                {
+                    // User belongs to a group that effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRGRP;
+                    writeBit = Interop.Sys.Permissions.S_IWGRP;
+                }
+                else
+                {
+                    // Others permissions
+                    readBit = Interop.Sys.Permissions.S_IROTH;
+                    writeBit = Interop.Sys.Permissions.S_IWOTH;
+                }
+    #endif
+                return ((_mainCache.Mode & (int)readBit) != 0 && // has read permission
+                    (_mainCache.Mode & (int)writeBit) == 0);     // but not write permission
             }
-            else if (_fileStatus.Gid == Interop.Sys.GetEGid())
+        }
+
+        internal bool HasSecondaryDirectoryFlag => HasDirectoryFlag(_secondaryCache);
```

Comment:
`HasSecondaryDirectoryFlag` does not tell me much. Is it checking if the symlink is a directory?

```suggestion
        internal bool IsSymlinkToDirectory => HasDirectoryFlag(_secondaryCache);
```

Reply:
Yes, it is checking if the symbolic link's _target_ is a directory.